### PR TITLE
add nodeSelector, affinity, and toleration support

### PR DESCRIPTION
**Pull request by pgvishnuram**
## Description

Add scheduling support for kibana index job and jetstream job

## Related Issues

- https://github.com/astronomer/issues/issues/6676
- https://github.com/astronomer/issues/issues/6684
- https://github.com/astronomer/astronomer/issues/2319

## Testing

QA should able to define nodeselector, tolleration, affinity for these pods 

## Merging

cherry-pick to all valid releases

### CODE REVIEW

**Review by Shubham-astro — APPROVED** [review]
Added scheduling for jetstream and kibana. looks good

**Review by rishkarajgi — DISMISSED** [review]
- PR description needs to be updated
- Add tests

File: charts/kibana/templates/kibana-default-index-cronjob.yaml [hunk 1]
```diff
@@ -32,6 +32,12 @@ spec:
           sidecar.istio.io/inject: "false"
         {{- end }}
       spec:
+        nodeSelector:
+{{ toYaml (default .Values.global.platformNodePool.nodeSelector .Values.nodeSelector) | indent 10 }}
+        affinity:
+{{ toYaml (default .Values.global.platformNodePool.affinity .Values.affinity) | indent 10 }}
+        tolerations:
+{{ toYaml (default .Values.global.platformNodePool.tolerations .Values.tolerations) | indent 10 }}
```
Comment (danielhoherd) [inline charts/kibana/templates/kibana-default-index-cronjob.yaml:40]:
```suggestion
        nodeSelector: {{ toYaml (default .Values.global.platformNodePool.nodeSelector .Values.nodeSelector) | nindent 10 }}
        affinity: {{ toYaml (default .Values.global.platformNodePool.affinity .Values.affinity) | nindent 10 }}
        tolerations: {{ toYaml (default .Values.global.platformNodePool.tolerations .Values.tolerations) | nindent 10 }}
```

File: charts/nats/templates/jetstream-job.yaml [hunk 2]
```diff
@@ -82,6 +82,12 @@ spec:
     spec:
       restartPolicy: Never
       serviceAccountName: {{ .Release.Name }}-jetstream-sa
+      nodeSelector:
+{{ toYaml (default .Values.global.platformNodePool.nodeSelector .Values.nodeSelector) | indent 8 }}
+      affinity:
+{{ toYaml (default .Values.global.platformNodePool.affinity .Values.affinity) | indent 8 }}
+      tolerations:
+{{ toYaml (default .Values.global.platformNodePool.tolerations .Values.tolerations) | indent 8 }}
```
Comment (danielhoherd) [inline charts/nats/templates/jetstream-job.yaml:90]:
```suggestion
      nodeSelector: {{ toYaml (default .Values.global.platformNodePool.nodeSelector .Values.nodeSelector) | nindent 8 }}
      affinity: {{ toYaml (default .Values.global.platformNodePool.affinity .Values.affinity) | nindent 8 }}
      tolerations: {{ toYaml (default .Values.global.platformNodePool.tolerations .Values.tolerations) | nindent 8 }}
```

**Review by danielhoherd — CHANGES_REQUESTED** [review]
This all look good so far.

We need test cases that show the scheduling constraints being customized.

File: charts/prometheus-postgres-exporter/templates/deployment.yaml [hunk 3]
```diff
@@ -37,6 +37,12 @@ spec:
 {{ toYaml .Values.annotations | indent 8 }}
 {{- end }}
     spec:
+      nodeSelector:
+{{ toYaml (default .Values.global.platformNodePool.nodeSelector .Values.nodeSelector) | indent 8 }}
```
Comment (danielhoherd) [inline charts/prometheus-postgres-exporter/templates/deployment.yaml:41]: IMHO we should nindent all of these so they are on a single line. it makes reading the template easier.

File: tests/chart_tests/test_kibana.py [hunk 4]
```diff
@@ -55,6 +55,9 @@ def test_kibana_index_defaults(self, kube_version):
         )
         common_kibana_cronjob_test(docs)
         doc = docs[0]
+        assert doc["spec"]["template"]["spec"]["nodeSelector"] == {}
```
Comment (danielhoherd) [inline tests/chart_tests/test_kibana.py:58]: We need a test case here to verify that custom nodeSelector, affinity, and tolerations work as expected.

File: tests/chart_tests/test_nats_jetstream.py [hunk 5]
```diff
@@ -68,6 +68,9 @@ def test_nats_statefulset_with_jetstream_and_tls(self, kube_version):
                 "keyFile": f"{jetStreamCertPrefix}-client/tls.key",
             },
         }
+        assert docs[7]["spec"]["template"]["spec"]["nodeSelector"] == {}
```
Comment (danielhoherd) [inline tests/chart_tests/test_nats_jetstream.py:71]: We need a test case here to verify that custom nodeSelector, affinity, and tolerations work as expected.

**Review by danielhoherd — CHANGES_REQUESTED** [review]
Looking good. Needs two more test cases, and one test case needs to be renamed and have its docstring updated.

File: charts/external-es-proxy/templates/external-es-proxy-deployment.yaml [hunk 6]
```diff
@@ -39,6 +39,12 @@ spec:
         iam.amazonaws.com/role: {{ .Values.global.customLogging.awsIAMRole }}
         {{- end }}
     spec:
+      nodeSelector:
+{{ toYaml (default .Values.global.platformNodePool.nodeSelector .Values.nodeSelector) | indent 12 }}
+      affinity:
```
Comment (danielhoherd) [inline charts/external-es-proxy/templates/external-es-proxy-deployment.yaml:44]: These could be nindented.

File: charts/elasticsearch/templates/curator/es-curator-cronjob.yaml [hunk 7]
```diff
@@ -36,6 +36,12 @@ spec:
             app: {{ template "elasticsearch.name" . }}-curator
             version: {{ .Chart.Version }}
         spec:
+          nodeSelector:
+{{ toYaml (default .Values.global.platformNodePool.nodeSelector .Values.nodeSelector) | indent 12 }}
```
Comment (danielhoherd) [inline charts/elasticsearch/templates/curator/es-curator-cronjob.yaml:40]: These could be nindented

File: tests/chart_tests/test_elasticsearch.py [hunk 8]
```diff
@@ -530,31 +531,100 @@ def test_elasticsearch_curator_cronjob_defaults(self, kube_version):
...
+
+    def test_elasticsearch_curator_cronjob_subchart_overrides(self, kube_version):
+        """Test ElasticSearch Curator cron job with nodeSelector, affinity, tolerations and config defaults"""
+        values = {
+            "elasticsearch": {
+                "curator": {
+                    "schedule": "0 45 * * *",
+                    "securityContext": {"runAsNonRoot": True},
+                },
+                "nodeSelector": {"role": "astroelasticsearch"},
```
Comment (danielhoherd) [inline tests/chart_tests/test_elasticsearch.py:587]: If we're going to put global_platform_node_pool_config in __init__.py, shouldn't we deepcopy that and update the fields we need to change?

File: tests/chart_tests/test_external_elasticsearch.py [hunk 9]
```diff
@@ -569,3 +569,108 @@ def test_external_elasticsearch_nginx_defaults_config(self, kube_version):
...
+        values = {
+            "global": {
+                "customLogging": {
+                    "enabled": True,
+                    "secret": secret,
+                    "host": "esdemo.example.com",
+                },
+            },
+            "external-es-proxy": {
+                "nodeSelector": {"role": "astroesproxy"},
```
Comment (danielhoherd) [inline tests/chart_tests/test_external_elasticsearch.py:637]: Same deepcopy question here

File: tests/chart_tests/test_kibana.py [hunk 10]
```diff
@@ -54,6 +55,80 @@ def test_kibana_index_defaults(self, kube_version):
...
+        assert doc["spec"]["template"]["spec"]["tolerations"] == values["global"]["platformNodePool"]["tolerations"]
+
+        assert "fluentd.*" in doc["spec"]["template"]["spec"]["containers"][0]["command"][2]
+
+    def test_kibana_index_defaults_with_subchart_overrides(self, kube_version):
+        """Test that kibana index cronjobs renders proper nodeSelector, affinity,
+        and tolerations with global config and index defaults."""
+        values = {
+            "kibana": {
+                "nodeSelector": {"role": "astrokibana"},
```
Comment (danielhoherd) [inline tests/chart_tests/test_kibana.py:91]: Same deepcopy question here

File: tests/chart_tests/test_elasticsearch.py [hunk 11]
```diff
@@ -530,31 +531,100 @@ def test_elasticsearch_curator_cronjob_defaults(self, kube_version):
...
+        assert len(docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]["tolerations"]) > 0
+        assert (
+            docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]["tolerations"]
+            == values["global"]["platformNodePool"]["tolerations"]
+        )
+        assert c_by_name["curator"]["command"] == ["/bin/sh", "-c"]
+        assert c_by_name["curator"]["args"] == [
+            "sleep 5; /usr/bin/curator --config /etc/config/config.yml /etc/config/action_file.yml; exit_code=$?; wget --timeout=5 -O- --post-data='not=used' http://127.0.0.1:15020/quitquitquit; exit $exit_code;"
+        ]
+        assert c_by_name["curator"]["securityContext"] == {"runAsNonRoot": True}
```
Comment (danielhoherd) [inline tests/chart_tests/test_elasticsearch.py:577]: move `c_by_name = ...` closer to where it is used
```suggestion
        assert docs[0]["kind"] == "CronJob"
        assert docs[0]["metadata"]["name"] == "release-name-elasticsearch-curator"
        assert docs[0]["spec"]["schedule"] == "0 45 * * *"
        assert docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]["nodeSelector"]["role"] == "astro"
        assert len(docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]["affinity"]) == 1
        assert len(docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]["tolerations"]) > 0
        assert (
            docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]["tolerations"]
            == values["global"]["platformNodePool"]["tolerations"]
        )
        c_by_name = get_cronjob_containerspec_by_name(docs[0])
        assert c_by_name["curator"]["command"] == ["/bin/sh", "-c"]
        assert c_by_name["curator"]["args"] == [
            "sleep 5; /usr/bin/curator --config /etc/config/config.yml /etc/config/action_file.yml; exit_code=$?; wget --timeout=5 -O- --post-data='not=used' http://127.0.0.1:15020/quitquitquit; exit $exit_code;"
        ]
        assert c_by_name["curator"]["securityContext"] == {"runAsNonRoot": True}
```

File: tests/chart_tests/test_elasticsearch.py [hunk 12]
```diff
@@ -530,31 +531,100 @@ def test_elasticsearch_curator_cronjob_defaults(self, kube_version):
...
+            kube_version=kube_version,
+            values=values,
             show_only=["charts/elasticsearch/templates/curator/es-curator-cronjob.yaml"],
         )
         assert len(docs) == 1
         c_by_name = get_cronjob_containerspec_by_name(docs[0])
         assert docs[0]["kind"] == "CronJob"
         assert docs[0]["metadata"]["name"] == "release-name-elasticsearch-curator"
         assert docs[0]["spec"]["schedule"] == "0 45 * * *"
+        assert docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]["nodeSelector"]["role"] == "astroelasticsearch"
```
Comment (danielhoherd) [inline tests/chart_tests/test_elasticsearch.py:624]: nit: With 4 uses, maybe `docs[0]["spec"]["jobTemplate"]["spec"]["template"]["spec"]` should get its own variable?

File: tests/chart_tests/test_external_elasticsearch.py [hunk 13]
```diff
@@ -569,3 +569,108 @@ def test_external_elasticsearch_nginx_defaults_config(self, kube_version):
...
+            show_only=[
+                "charts/external-es-proxy/templates/external-es-proxy-deployment.yaml",
+            ],
+        )
+
+        assert len(docs) == 1
+        doc = docs[0]
+        assert doc["spec"]["template"]["spec"]["nodeSelector"] == {}
+        assert doc["spec"]["template"]["spec"]["affinity"] == {}
+        assert doc["spec"]["template"]["spec"]["tolerations"] == []
```
Comment (danielhoherd) [inline tests/chart_tests/test_external_elasticsearch.py:596]:
```suggestion
        spec = docs[0]["spec"]["template"]["spec"]
        assert spec["nodeSelector"] == {}
        assert spec["affinity"] == {}
        assert spec["tolerations"] == []
```

File: tests/chart_tests/test_kibana.py [hunk 14]
```diff
@@ -54,6 +55,80 @@ def test_kibana_index_defaults(self, kube_version):
...
+                "charts/kibana/templates/kibana-default-index-cronjob.yaml",
+            ],
+        )
+        common_kibana_cronjob_test(docs)
+        doc = docs[0]
+        assert len(doc["spec"]["template"]["spec"]["nodeSelector"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["tolerations"]) > 0
+        assert doc["spec"]["template"]["spec"]["tolerations"] == values["global"]["platformNodePool"]["tolerations"]
+
+        assert "fluentd.*" in doc["spec"]["template"]["spec"]["containers"][0]["command"][2]
```
Comment (danielhoherd) [inline tests/chart_tests/test_kibana.py:84]: Maybe this?

```suggestion
        spec = docs[0]["spec"]["template"]["spec"]
        assert len(spec["nodeSelector"]) == 1
        assert len(spec["tolerations"]) > 0
        assert spec["tolerations"] == values["global"]["platformNodePool"]["tolerations"]

        assert "fluentd.*" in spec["containers"][0]["command"][2]
```

File: tests/chart_tests/test_prometheus_blackbox_exporter.py [hunk 15]
```diff
@@ -95,3 +105,66 @@ def test_prometheus_blackbox_exporter_deployment_custom_security_context(self, k
...
+        docs = render_chart(
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-blackbox-exporter/templates/deployment.yaml"],
+        )
+        doc = docs[0]
+        assert len(doc["spec"]["template"]["spec"]["nodeSelector"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["affinity"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["tolerations"]) > 0
+        doc["spec"]["template"]["spec"]["nodeSelector"] == "astro-prometheus-blackbox-exporter"
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_blackbox_exporter.py:142]: Is this supposed to be an assignment, not an assertion?

File: tests/chart_tests/test_prometheus_postgres_exporter.py [hunk 16]
```diff
@@ -48,3 +48,72 @@ def test_prometheus_postgres_exporter_service_enabled(self, kube_version):
...
+        docs = render_chart(
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-blackbox-exporter/templates/deployment.yaml"],
+        )
+        assert len(docs) == 1
+        doc = docs[0]
+        assert len(doc["spec"]["template"]["spec"]["nodeSelector"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["tolerations"]) > 0
+        assert doc["spec"]["template"]["spec"]["tolerations"] == values["global"]["platformNodePool"]["tolerations"]
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_postgres_exporter.py:73]:
```suggestion
        spec = docs[0]["spec"]["template"]["spec"]
        assert len(spec["nodeSelector"]) == 1
        assert len(spec["tolerations"]) > 0
        assert spec["tolerations"] == values["global"]["platformNodePool"]["tolerations"]
```

File: tests/chart_tests/test_prometheus_postgres_exporter.py [hunk 17]
```diff
@@ -48,3 +48,72 @@ def test_prometheus_postgres_exporter_service_enabled(self, kube_version):
...
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-postgres-exporter/templates/deployment.yaml"],
+        )
+        assert len(docs) == 1
+        doc = docs[0]
+        assert len(doc["spec"]["template"]["spec"]["nodeSelector"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["affinity"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["tolerations"]) > 0
+        doc["spec"]["template"]["spec"]["nodeSelector"] == "astro-prometheus-postgres-exporter"
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_postgres_exporter.py:94]: This should be an assertion, right?

File: tests/chart_tests/test_prometheus_postgres_exporter.py [hunk 18]
```diff
@@ -48,3 +48,72 @@ def test_prometheus_postgres_exporter_service_enabled(self, kube_version):
...
+                ],
+            },
+        }
+        docs = render_chart(
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-postgres-exporter/templates/deployment.yaml"],
+        )
+        assert len(docs) == 1
+        doc = docs[0]
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_postgres_exporter.py:90]: We could do `doc = docs[0]["spec"]["template"]["spec"]` here

File: tests/chart_tests/test_prometheus_postgres_exporter.py [hunk 19]
```diff
@@ -48,3 +48,72 @@ def test_prometheus_postgres_exporter_service_enabled(self, kube_version):
...
+                "prometheusPostgresExporterEnabled": True,
+                "platformNodePool": global_platform_node_pool_config,
+            }
+        }
+        docs = render_chart(
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-blackbox-exporter/templates/deployment.yaml"],
+        )
+        assert len(docs) == 1
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_postgres_exporter.py:69]: We could do `doc = docs[0]["spec"]["template"]["spec"]` here

File: tests/chart_tests/test_prometheus_blackbox_exporter.py [hunk 20]
```diff
@@ -95,3 +105,66 @@ def test_prometheus_blackbox_exporter_deployment_custom_security_context(self, k
...
+                    }
+                ],
+            }
+        }
+        docs = render_chart(
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-blackbox-exporter/templates/deployment.yaml"],
+        )
+        doc = docs[0]
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_blackbox_exporter.py:138]: We could do `doc = docs[0]["spec"]["template"]["spec"]` here

File: tests/chart_tests/test_prometheus_postgres_exporter.py [hunk 21]
```diff
@@ -48,3 +48,49 @@ def test_prometheus_postgres_exporter_service_enabled(self, kube_version):
...
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-postgres-exporter/templates/deployment.yaml"],
+        )
+        assert len(docs) == 1
+        spec = docs[0]["spec"]["template"]["spec"]
+        assert len(spec["nodeSelector"]) == 1
+        assert len(spec["affinity"]) == 1
+        assert len(spec["tolerations"]) > 0
+        spec["nodeSelector"] == values["prometheus-postgres-exporter"]["nodeSelector"]
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_postgres_exporter.py:95]: This seems wrong. Why is this not an assertion if it is a comparison?

File: tests/chart_tests/test_prometheus_blackbox_exporter.py [hunk 22]
```diff
@@ -95,3 +105,39 @@ def test_prometheus_blackbox_exporter_deployment_custom_security_context(self, k
...
+        docs = render_chart(
+            kube_version=kube_version,
+            values=values,
+            show_only=["charts/prometheus-blackbox-exporter/templates/deployment.yaml"],
+        )
+        spec = docs[0]["spec"]["template"]["spec"]
+        assert len(spec["nodeSelector"]) == 1
+        assert len(spec["affinity"]) == 1
+        assert len(spec["tolerations"]) > 0
+        spec["nodeSelector"] == values["prometheus-blackbox-exporter"]["nodeSelector"]
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_blackbox_exporter.py:142]: This seems wrong. Why is this not an assertion if it is a comparison?

File: tests/chart_tests/test_prometheus_blackbox_exporter.py [hunk 23]
```diff
@@ -67,26 +77,26 @@ def test_prometheus_blackbox_exporter_deployment_custom_resources(self, kube_ver
...
             kube_version=kube_version,
             values={
                 "prometheus-blackbox-exporter": {"securityContext": {"runAsUser": 1000}},
             },
             show_only=["charts/prometheus-blackbox-exporter/templates/deployment.yaml"],
-        )[0]
-
+        )
+        common_blackbox_exporter_tests(docs)
+        doc = docs[0]
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_blackbox_exporter.py:99]: We dont't need to assign `doc` if we're only going to use it once on the next line. Just use `docs[0]` on the next line.

File: tests/chart_tests/test_prometheus_blackbox_exporter.py [hunk 24]
```diff
@@ -67,26 +77,26 @@ def test_prometheus_blackbox_exporter_deployment_custom_resources(self, kube_ver
...
             },
             show_only=["charts/prometheus-blackbox-exporter/templates/deployment.yaml"],
-        )[0]
-
-        assert doc["kind"] == "Deployment"
-        assert doc["metadata"]["name"] == "release-name-prometheus-blackbox-exporter"
+        )
 
+        common_blackbox_exporter_tests(docs)
+        doc = docs[0]
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_blackbox_exporter.py:83]: We dont't need to assign `doc` if we're only going to use it once on the next line. Just use `docs[0]` on the next line.

File: tests/chart_tests/test_prometheus_blackbox_exporter.py [hunk 25]
```diff
@@ -36,12 +47,11 @@ def test_prometheus_blackbox_exporter_deployment_defaults(self, kube_version):
...
         )
 
-        assert len(docs) == 1
+        common_blackbox_exporter_tests(docs)
         doc = docs[0]
-        assert doc["kind"] == "Deployment"
-        assert doc["metadata"]["name"] == "release-name-prometheus-blackbox-exporter"
-        assert doc["spec"]["selector"]["matchLabels"]["component"] == "blackbox-exporter"
-        assert doc["spec"]["template"]["metadata"]["labels"]["app"] == "prometheus-blackbox-exporter"
+        assert doc["spec"]["template"]["spec"]["nodeSelector"] == {}
```
Comment (danielhoherd) [inline tests/chart_tests/test_prometheus_blackbox_exporter.py:52]: We could do the spec assignment here to make this shorter.

File: tests/chart_tests/test_kibana.py [hunk 26]
```diff
@@ -54,6 +55,53 @@ def test_kibana_index_defaults(self, kube_version):
...
+            show_only=[
+                "charts/kibana/templates/kibana-default-index-cronjob.yaml",
+            ],
+        )
+        common_kibana_cronjob_test(docs)
+        doc = docs[0]
+        assert len(doc["spec"]["template"]["spec"]["nodeSelector"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["affinity"]) == 1
+        assert len(doc["spec"]["template"]["spec"]["tolerations"]) > 0
+        doc["spec"]["template"]["spec"]["nodeSelector"] == "astrokibana"
```
Comment (danielhoherd) [inline tests/chart_tests/test_kibana.py:103]: Why is this comparison not an assertion?

File: tests/chart_tests/test_kibana.py [hunk 27]
```diff
@@ -54,6 +55,53 @@ def test_kibana_index_defaults(self, kube_version):
...
+        values = {"kibana": global_platform_node_pool_config}
+        docs = render_chart(
+            kube_version=kube_version,
+            values=values,
+            show_only=[
+                "charts/kibana/templates/kibana-default-index-cronjob.yaml",
+            ],
+        )
+        common_kibana_cronjob_test(docs)
+        doc = docs[0]
```
Comment (danielhoherd) [inline tests/chart_tests/test_kibana.py:99]: We could do the `spec = docs[0]['spec']['template']['spec']` thing here.

File: tests/chart_tests/test_kibana.py [hunk 28]
```diff
@@ -54,6 +55,53 @@ def test_kibana_index_defaults(self, kube_version):
         )
         common_kibana_cronjob_test(docs)
         doc = docs[0]
+        assert doc["spec"]["template"]["spec"]["nodeSelector"] == {}
```
Comment (danielhoherd) [inline tests/chart_tests/test_kibana.py:58]: We could do the `spec = docs[0]['spec']['template']['spec']` thing here.

File: tests/chart_tests/test_external_elasticsearch.py [hunk 29]
```diff
@@ -569,3 +569,83 @@ def test_external_elasticsearch_nginx_defaults_config(self, kube_version):
...
+
+        assert len(docs) == 1
+        spec = docs[0]["spec"]["template"]["spec"]
+        assert len(spec["nodeSelector"]) == 1
+        assert len(spec["tolerations"]) > 0
+        assert spec["tolerations"] == values["global"]["platformNodePool"]["tolerations"]
+
+    def test_external_elasticsearch_nginx_deployment_with_subchart_overrides(self, kube_version):
+        """Test that External ElasticSearch renders proper nodeSelector, affinity,
+        and tolerations with global config and nginx defaults"""
```
Comment (danielhoherd) [inline tests/chart_tests/test_external_elasticsearch.py:627]:
```suggestion
        and tolerations with global config and nginx overrides."""
```

File: tests/chart_tests/test_external_elasticsearch.py [hunk 30]
```diff
@@ -569,3 +569,83 @@ def test_external_elasticsearch_nginx_defaults_config(self, kube_version):
...
+
+        assert len(docs) == 1
+        spec = docs[0]["spec"]["template"]["spec"]
+        assert spec["nodeSelector"] == {}
+        assert spec["affinity"] == {}
+        assert spec["tolerations"] == []
+
+    def test_external_elasticsearch_nginx_deployment_global_platformnodepool_overrides(self, kube_version):
+        """Test that External ElasticSearch renders proper nodeSelector, affinity,
+        and tolerations with global config and nginx defaults"""
```
Comment (danielhoherd) [inline tests/chart_tests/test_external_elasticsearch.py:600]:
```suggestion
        and tolerations with global config and nginx overrides"""
```

File: tests/chart_tests/test_elasticsearch.py [hunk 31]
```diff
@@ -530,31 +531,79 @@ def test_elasticsearch_curator_cronjob_defaults(self, kube_version):
...
+        assert spec["tolerations"] == values["global"]["platformNodePool"]["tolerations"]
+        c_by_name = get_cronjob_containerspec_by_name(docs[0])
+        assert c_by_name["curator"]["command"] == ["/bin/sh", "-c"]
+        assert c_by_name["curator"]["args"] == [
+            "sleep 5; /usr/bin/curator --config /etc/config/config.yml /etc/config/action_file.yml; exit_code=$?; wget --timeout=5 -O- --post-data='not=used' http://127.0.0.1:15020/quitquitquit; exit $exit_code;"
+        ]
+        assert c_by_name["curator"]["securityContext"] == {"runAsNonRoot": True}
+
+    def test_elasticsearch_curator_cronjob_subchart_overrides(self, kube_version):
+        """Test ElasticSearch Curator cron job with nodeSelector, affinity, tolerations and config defaults"""
```
Comment (danielhoherd) [inline tests/chart_tests/test_elasticsearch.py:579]:
```suggestion
        """Test ElasticSearch Curator cron job with nodeSelector, affinity, tolerations and config overrides."""
```

File: tests/chart_tests/test_elasticsearch.py [hunk 32]
```diff
@@ -530,31 +531,79 @@ def test_elasticsearch_curator_cronjob_defaults(self, kube_version):
...
+        assert spec["tolerations"] == []
         assert c_by_name["curator"]["command"] == ["/bin/sh", "-c"]
         assert c_by_name["curator"]["args"] == [
             "sleep 5; /usr/bin/curator --config /etc/config/config.yml /etc/config/action_file.yml; exit_code=$?; wget --timeout=5 -O- --post-data='not=used' http://127.0.0.1:15020/quitquitquit; exit $exit_code;"
         ]
         assert c_by_name["curator"]["securityContext"] == {}
 
     def test_elasticsearch_curator_cronjob_overrides(self, kube_version):
-        """Test ElasticSearch Curator cron job with defaults"""
+        """Test ElasticSearch Curator cron job with nodeSelector, affinity, tolerations and config defaults"""
```
Comment (danielhoherd) [inline tests/chart_tests/test_elasticsearch.py:545]:
```suggestion
        """Test ElasticSearch Curator cron job with nodeSelector, affinity, tolerations and config overrides."""
```

File: tests/chart_tests/test_stan_sts.py [hunk 33]
```diff
@@ -37,9 +37,10 @@ def test_stan_statefulset_defaults(self, kube_version):
         }
 
         assert all(c["securityContext"] == {"runAsNonRoot": True} for c in c_by_name.values())
-        assert doc["spec"]["template"]["spec"]["nodeSelector"] == {}
-        assert doc["spec"]["template"]["spec"]["affinity"] == {}
-        assert doc["spec"]["template"]["spec"]["tolerations"] == []
+        spec = doc["spec"]["template"]["spec"]
+        assert spec["nodeSelector"] == {}
+        assert spec["affinity"] == {}
+        assert spec["tolerations"] == []
```
Comment (danielhoherd) [inline tests/chart_tests/test_stan_sts.py:43]: We should replace the values on line 96 in this file with global_platform_node_pool_config

**Review by danielhoherd — APPROVED** [review]
Lots of new code, but after a bunch of back and forth I think this looks great.

I left one more optional improvement to replace prior test values with global_platform_node_pool_config

**Pull request merged**